nengo/builder/optimizer.py

```diff
@@ -0,0 +1,969 @@
+                if can_merge:
+                    merge.append(op2)
+                elif self._check_sequential(merge[-1], op2):
+                    # If this check is true the view of op2 does not
+                    # immediately follow the view of the operators being
+                    # merged. Because we iterate over the operators sorted by
+                    # view offset there will be a gap between the end of the
+                    # operators being merged and all remaining operators to
+                    # loop over. With such a gap a merge is never possible and
+                    # we can cut the loop short.
+                    break
+
+            if len(merge) > 1:
+                self._merge(merge, op_replacements, sig_replacements)
+
+        return op_replacements, sig_replacements
+
+    def _merge(self, merge, op_replacements, sig_replacements):
+        """Merges the operators in `merge` and updates `op_replacements` and
+        `sig_replacements` with the required replacements of operators and
+        signals.
+
+        It will also update the internal list of merged operators to prevent
+        further merges on the same operators before all required operators and
+        signals have been replaced.
+        """
+        merged_op, merged_sig = self.merge_ops(merge)
+        self._merged.update(merge)
+        for op in merge:
+            op_replacements[op] = merged_op
+            # Mark all operators referencing the same signals as merged
+            # (even though they are not) to prevent them from getting
+            # merged before their signals have been updated.
+            for s in op.all_signals:
+                self._merged.update(self._sig2op[s])
+        sig_replacements.update(merged_sig)
+
+    def _update_dg(self, op_replacements):
+        """Updates the dependency graph.
+
+        This involves doing all operator replacements in `op_replacements` and
+        ensuring that dependencies are preserved.
+        """
+        dg = {}
+        for op in op_replacements:
+            if op_replacements[op] not in dg:
+                dg[op_replacements[op]] = set()
+            for x in self.dg[op]:
+                dg[op_replacements[op]].add(op_replacements[x])
+        self.dg.clear()
+        self.dg.update(dg)
+
+    @staticmethod
+    def _map_onto_op_signals(fn, ops):
+        """Maps function `fn` on all signals of all operators in the sequence
+        `ops`.
+        """
+        for op in ops:
+            for key in dir(op):
+                sig = getattr(op, key)
+                if isinstance(sig, Signal):
+                    setattr(op, key, fn(sig))
+
+            op.sets = [fn(s) for s in op.sets]
+            op.incs = [fn(s) for s in op.incs]
+            op.reads = [fn(s) for s in op.reads]
+            op.updates = [fn(s) for s in op.updates]
+
+    @classmethod
+    def _get_sig_view_replacements(cls, ops, sig_replacements):
```

Comment:
I'm confused because I think a lot of stuff is being passed by reference and then the references are being modified. Is this function supposed to modify the `ops`?